tests/tests_integration/test_api/test_events.py

```diff
@@ -109,3 +112,213 @@ def test_delete_with_nonexisting(self, cognite_client):
+            external_id="test_upsert_all_preexisting:new",
+            type="test__py__sdk",
+            start_time=0,
+            end_time=1,
+            subtype="mySubType1",
+        )
+
+        try:
+            _ = cognite_client.events.create(new_event)
+
+            # Act
+            res = cognite_client.events.upsert(new_event)
+
+            # Assert
+            assert isinstance(res, Event)
+            assert new_event.external_id == res.external_id
+        finally:
+            cognite_client.events.delete(external_id=new_event.external_id, ignore_unknown_ids=True)
+
+    def test_upsert_without_external_id(self, cognite_client: CogniteClient) -> None:
+        # Arrange
+        new_event = Event(
+            external_id="test_upsert_without_external_id:new",
+            type="test__py__sdk",
+            start_time=0,
+            end_time=1,
+            subtype="mySubType1",
+        )
+        existing = Event(
+            external_id="test_upsert_without_external_id:existing",
+            type="test__py__sdk",
+            start_time=0,
+            end_time=1,
+            subtype="mySubType2",
+        )
+        existing_update = Event._load(existing.dump(camel_case=True))
+        existing_update.subtype = "mySubType1"
+
+        try:
+            created = cognite_client.events.create(existing)
+            existing_update.external_id = None
+            existing_update.id = created.id
+
+            # Act
+            res = cognite_client.events.upsert([new_event, existing_update])
+
+            # Assert
+            assert len(res) == 2
+            assert new_event.external_id == res[0].external_id
+            assert existing.external_id == res[1].external_id
+            assert new_event.subtype == res[0].subtype
+            assert existing_update.subtype == res[1].subtype
+        finally:
+            cognite_client.events.delete(
+                external_id=[new_event.external_id, existing.external_id], ignore_unknown_ids=True
+            )
+
+    def test_upsert_split_into_multiple_tasks(self, cognite_client: CogniteClient, monkeypatch: MonkeyPatch) -> None:
+        # Arrange
+        new_event = Event(
+            external_id="test_upsert_split_into_multiple_tasks:new",
+            type="test__py__sdk",
+            start_time=0,
+            end_time=1,
+            subtype="mySubType1",
+        )
+        preexisting = Event(
+            external_id="test_upsert_split_into_multiple_tasks:preexisting",
+            type="test__py__sdk",
+            start_time=0,
+            end_time=1,
+            subtype="mySubType2",
+        )
+        preexisting_update = Event._load(preexisting.dump(camel_case=True))
+        preexisting_update.subtype = "mySubType1"
+
+        try:
+            created_existing = cognite_client.events.create(preexisting)
+            assert created_existing is not None
+            monkeypatch.setattr(cognite_client.events, "_UPDATE_LIMIT", 1)
```

Comment:
I would add an assert that verifies that you send multiple API requests just in case we rewrite the current thread-pool-executor-monster 😅

Reply:
Do you have a simple code snippet for doing that?

Reply:
Yup, check out the `post_spy`, which works like a "man in the middle" for you:
```py
@pytest.fixture
def post_spy(cognite_client):
    dps_api = cognite_client.time_series.data
    with patch.object(dps_api, "_post", wraps=dps_api._post):
        yield
```